### ISSUE TITLE
Using Outbox results in multi-subscriber events being processed by only one subscriber

### ISSUE THREAD

**Issue by indualagarsamy** [Particular/NServiceBus.NHibernate#117]
### Who's affected

To be affected by this bug, you must be using the Outbox feature for message deduplication with NHibernate persistence, and publish events to multiple subscribers.

### Symptoms

An event is published for multiple subscribers, but only one subscriber successfully receives and processes the message. Other subscribers may receive the event and attempt to process it, but are unable to do so, logging an exception similar to the following:

```
2015-10-13 09:42:47.679 INFO  NServiceBus.Unicast.Transport.TransportReceiver Failed to process message
NHibernate.Exceptions.GenericADOException: could not insert: [NServiceBus.Outbox.NHibernate.OutboxRecord][SQL: INSERT INTO OutboxRecord (MessageId, Dispatched, DispatchedAt, TransportOperations) VALUES (?, ?, ?, ?); select SCOPE_IDENTITY()] ---> System.Data.SqlClient.SqlException: Violation of UNIQUE KEY constraint 'UQ__OutboxRe__C87C0C9D3A7D137D'. Cannot insert duplicate key in object 'dbo.OutboxRecord'. The duplicate key value is (ed872e70-91ca-4c17-992a-a53000a00e98).
The statement has been terminated.
```

This occurs because of improper sharing of Outbox records. There is a unique key collision, causing the second endpoint to fail. When the second endpoint attempts to retry, it sees an existing Outbox record and does not attempt to reprocess the message, believing that it has already been successfully processed.

---

### Root cause

The outbox table is shared between all endpoints but the table does not contain a column storing the endpoint that owns the record. Because of that, if two endpoints get the same message, they both attempt to insert an outbox record with same ID, one of them fails. It tries again, sees that there is already a record in the DB so it skips the handler (drops a message).

### Plan of attack

 - [x] Modify outbox to store records with stringified composite id,  `{EndpointName}/{MessageId}` using existing schema.
 - [x] Modify outbox cleanup to use straight SQL - endpoints will clean up each other's expired items but at least they won't fail.
 - [x] Create Raven query for ServiceControl -maint mode that can be executed in Raven Studio to help determine if customer has a problem.
 - [ ] Doco
 - [ ] Release & Announcement
 - Post Release
   - [ ] Verify if the same problem exists in RavenDB outbox persistence, raise new card to fix if necessary
   - [x] Create issue for Core V6 + NH V7 to move to table per endpoint storage #119 

### Impact assessment

Customer must be using Outbox + NHibernate +  publishing events, with multiple subscribers. 

While this can occur on any transport using the Outbox, it's unlikely that it is (at least in widespread use) because use of Outbox on MSMQ, for example, is atypical. Given this combination of factors and lack of complaints, this number can't be high. That said, when it occurs, it is effective message loss. Therefore, I say Impact: Medium. Size is a couple edits that should be fairly quick, I'd say Small.

### Customer communication

We have discovered a bug with the NHibernate implementation of the Outbox feature. Although the intersection of features necessary for this bug to present itself (NHibernate + Outbox + publishing multi-subscriber events) makes its overall impact fairly small, we are still taking this quite seriously as, for those customers that are affected, it effectively leads to message loss.

### Original issue

When outbox is enabled and using a broker transport like RabbitMQ, when a publisher publishes an event and multiple subscribers are subscribed to it, only one of the subscribers is able to receive this event successfully. 

The other subscriber receives the following exception:
```
2015-09-11 16:24:36.671 INFO  NServiceBus.Unicast.Transport.TransportReceiver Fa
iled to process message
NHibernate.Exceptions.GenericADOException: could not insert: [NServiceBus.Outbox
.NHibernate.OutboxRecord][SQL: INSERT INTO OutboxRecord (MessageId, Dispatched,
DispatchedAt, TransportOperations) VALUES (?, ?, ?, ?); select SCOPE_IDENTITY()]
 ---> System.Data.SqlClient.SqlException: Violation of UNIQUE KEY constraint 'UQ
__OutboxRe__C87C0C9DEC77C050'. Cannot insert duplicate key in object 'dbo.Outbox
Record'. The duplicate key value is (2c8c8b5b-4e4b-4486-b4d4-a510010e6b71).
The statement has been terminated.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolea
n breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception
, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObj
ect stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand
 cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler,
TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, Run
Behavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBe
havior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 time
out, Task& task, Boolean asyncWrite, SqlDataReader ds, Boolean describeParameter
EncryptionRequest)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehav
ior, RunBehavior runBehavior, Boolean returnStream, String method, TaskCompletio
nSource`1 completion, Int32 timeout, Task& task, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehav
ior, RunBehavior runBehavior, Boolean returnStream, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, S
tring method)
   at System.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behav
ior)
   at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader()
   at NHibernate.AdoNet.AbstractBatcher.ExecuteReader(IDbCommand cmd)
   at NHibernate.Id.IdentityGenerator.InsertSelectDelegate.ExecuteAndExtract(IDb
Command insert, ISessionImplementor session)
   at NHibernate.Id.Insert.AbstractReturningDelegate.PerformInsert(SqlCommandInf
o insertSQL, ISessionImplementor session, IBinder binder)
   --- End of inner exception stack trace ---
   at NHibernate.Id.Insert.AbstractReturningDelegate.PerformInsert(SqlCommandInf
o insertSQL, ISessionImplementor session, IBinder binder)
   at NHibernate.Persister.Entity.AbstractEntityPersister.Insert(Object[] fields
, Boolean[] notNull, SqlCommandInfo sql, Object obj, ISessionImplementor session
)
   at NHibernate.Persister.Entity.AbstractEntityPersister.Insert(Object[] fields
, Object obj, ISessionImplementor session)
   at NHibernate.Action.EntityIdentityInsertAction.Execute()
   at NHibernate.Engine.ActionQueue.Execute(IExecutable executable)
   at NHibernate.Event.Default.AbstractSaveEventListener.PerformSaveOrReplicate(
Object entity, EntityKey key, IEntityPersister persister, Boolean useIdentityCol
umn, Object anything, IEventSource source, Boolean requiresImmediateIdAccess)
   at NHibernate.Event.Default.AbstractSaveEventListener.PerformSave(Object enti
ty, Object id, IEntityPersister persister, Boolean useIdentityColumn, Object any
thing, IEventSource source, Boolean requiresImmediateIdAccess)
   at NHibernate.Event.Default.AbstractSaveEventListener.SaveWithGeneratedId(Obj
ect entity, String entityName, Object anything, IEventSource source, Boolean req
uiresImmediateIdAccess)
   at NHibernate.Event.Default.DefaultSaveOrUpdateEventListener.SaveWithGenerate
dOrRequestedId(SaveOrUpdateEvent event)
   at NHibernate.Event.Default.DefaultSaveOrUpdateEventListener.EntityIsTransien
t(SaveOrUpdateEvent event)
   at NHibernate.Event.Default.DefaultSaveEventListener.PerformSaveOrUpdate(Save
OrUpdateEvent event)
   at NHibernate.Event.Default.DefaultSaveOrUpdateEventListener.OnSaveOrUpdate(S
aveOrUpdateEvent event)
   at NHibernate.Impl.SessionImpl.FireSave(SaveOrUpdateEvent event)
   at NHibernate.Impl.SessionImpl.Save(Object obj)
   at NServiceBus.Outbox.OutboxPersister.<>c__DisplayClass7.<Store>b__5(ISession
 x) in c:\BuildAgent\work\5135de308b2f3016\src\NServiceBus.NHibernate\Outbox\Out
boxPersister.cs:line 62
   at NServiceBus.Persistence.NHibernate.SharedConnectionStorageSessionProvider.
ExecuteInTransaction(Action`1 operation) in c:\BuildAgent\work\5135de308b2f3016\
src\NServiceBus.NHibernate\SharedSession\SharedConnectionStorageSessionProvider.
cs:line 31
   at NServiceBus.Outbox.OutboxPersister.Store(String messageId, IEnumerable`1 t
ransportOperations) in c:\BuildAgent\work\5135de308b2f3016\src\NServiceBus.NHibe
rnate\Outbox\OutboxPersister.cs:line 68
   at NServiceBus.OutboxRecordBehavior.Invoke(IncomingContext context, Action ne
xt) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Outbox\OutboxRec
ordBehavior.cs:line 24
   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.<>c__DisplayClass5.<InvokeNext>b__2() in c:\Bu
ildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:li
ne 95
   at NServiceBus.Persistence.NHibernate.OpenSessionBehavior.InnerInvoke(Behavio
rContext context, Action next, Func`1 connectionRetriever) in c:\BuildAgent\work
\5135de308b2f3016\src\NServiceBus.NHibernate\SharedSession\OpenSessionBehavior.c
s:line 83
   at NServiceBus.Persistence.NHibernate.OpenSessionBehavior.Invoke(IncomingCont
ext context, Action next) in c:\BuildAgent\work\5135de308b2f3016\src\NServiceBus
.NHibernate\SharedSession\OpenSessionBehavior.cs:line 37
   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.<>c__DisplayClass5.<InvokeNext>b__2() in c:\Bu
ildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:li
ne 95
   at NServiceBus.SubscriptionReceiverBehavior.Invoke(IncomingContext context, A
ction next) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\
Subscriptions\MessageDrivenSubscriptions\SubscriptionReceiverBehavior.cs:line 32

   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.<>c__DisplayClass5.<InvokeNext>b__2() in c:\Bu
ildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:li
ne 95
   at NServiceBus.UnitOfWorkBehavior.Invoke(IncomingContext context, Action next
) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\UnitOfWork\UnitOfW
orkBehavior.cs:line 42
   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.<>c__DisplayClass5.<InvokeNext>b__2() in c:\Bu
ildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:li
ne 95
   at NServiceBus.Transports.RabbitMQ.OpenPublishChannelBehavior.Invoke(Incoming
Context context, Action next) in c:\BuildAgent\work\ef98ad7376e3379a\src\NServic
eBus.RabbitMQ\OpenPublishChannelBehavior.cs:line 19
   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.<>c__DisplayClass5.<InvokeNext>b__2() in c:\Bu
ildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:li
ne 95
   at NServiceBus.ProcessingStatisticsBehavior.Invoke(IncomingContext context, A
ction next) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Monitori
ng\ProcessingStatisticsBehavior.cs:line 23
   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.<>c__DisplayClass5.<InvokeNext>b__2() in c:\Bu
ildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:li
ne 95
   at NServiceBus.AuditBehavior.Invoke(IncomingContext context, Action next) in
c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Audit\AuditBehavior.cs:
line 22
   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.<>c__DisplayClass5.<InvokeNext>b__2() in c:\Bu
ildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:li
ne 95
   at NServiceBus.OutboxDeduplicationBehavior.Invoke(IncomingContext context, Ac
tion next) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Outbox\Ou
tboxDeduplicationBehavior.cs:line 40
   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.<>c__DisplayClass5.<InvokeNext>b__2() in c:\Bu
ildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:li
ne 95
   at NServiceBus.Persistence.NHibernate.OpenSqlConnectionBehavior.Invoke(Incomi
ngContext context, Action next) in c:\BuildAgent\work\5135de308b2f3016\src\NServ
iceBus.NHibernate\SharedSession\OpenSqlConnectionBehavior.cs:line 32
   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.<>c__DisplayClass5.<InvokeNext>b__2() in c:\Bu
ildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:li
ne 95
   at NServiceBus.ChildContainerBehavior.Invoke(IncomingContext context, Action
next) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\Behavi
ors\ChildContainerBehavior.cs:line 17
   at NServiceBus.BehaviorChain`1.InvokeNext(T context) in c:\BuildAgent\work\32
06e2123f54fce4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 107
   at NServiceBus.BehaviorChain`1.Invoke() in c:\BuildAgent\work\3206e2123f54fce
4\src\NServiceBus.Core\Pipeline\BehaviorChain.cs:line 52
   at NServiceBus.Pipeline.PipelineExecutor.Execute[T](BehaviorChain`1 pipelineA
ction, T context) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pi
peline\PipelineExecutor.cs:line 129
   at NServiceBus.Pipeline.PipelineExecutor.InvokePipeline[TContext](IEnumerable
`1 behaviors, TContext context) in c:\BuildAgent\work\3206e2123f54fce4\src\NServ
iceBus.Core\Pipeline\PipelineExecutor.cs:line 86
   at NServiceBus.Pipeline.PipelineExecutor.InvokeReceivePhysicalMessagePipeline
() in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\Pipeline
Executor.cs:line 103
   at NServiceBus.Unicast.UnicastBus.TransportMessageReceived(Object sender, Tra
nsportMessageReceivedEventArgs e) in c:\BuildAgent\work\3206e2123f54fce4\src\NSe
rviceBus.Core\Unicast\UnicastBus.cs:line 827
   at NServiceBus.Unicast.Transport.TransportReceiver.OnTransportMessageReceived
(TransportMessage msg) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Co
re\Unicast\Transport\TransportReceiver.cs:line 413
   at NServiceBus.Unicast.Transport.TransportReceiver.ProcessMessage(TransportMe
ssage message) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unica
st\Transport\TransportReceiver.cs:line 344
   at NServiceBus.Unicast.Transport.TransportReceiver.TryProcess(TransportMessag
e message) in c:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\T
ransport\TransportReceiver.cs:line 230
   at NServiceBus.Transports.RabbitMQ.RabbitMqDequeueStrategy.ConsumeMessages(Ob
ject state) in c:\BuildAgent\work\ef98ad7376e3379a\src\NServiceBus.RabbitMQ\Rabb
itMqDequeueStrategy.cs:line 193

```

Related desk case: https://nservicebus.desk.com/agent/case/13177

**Comment by SzymonPobiega** [comment]
Looking at V6 codebase, in V6 the problem would also manifest itself with non-broker transports. We decided with @andreasohlund to align non-broker transports in V6 with brokers and use the same message ID for outgoing events. 

The problem is, out outbox persistence is keyed per message ID only. Endpoints cannot share the same database for outbox (like they can for subscriptions and timeouts). Not sure why it has been implemented that way. @andreasohlund maybe?

**Comment by andreasohlund** [comment]
I can't remeber that we had any reasons to do it this way :(

+1 For fixing it

**Comment by SzymonPobiega** [comment]
@indualagarsamy @andreasohlund @johnsimons @SeanFeldman I've added plan of attack.

**Comment by SeanFeldman** [comment]
@SzymonPobiega why not to update `OutboxRecord` to include endpoint instance ID rather than creating a table per endpoint instance instead?
/cc @colin-higgins

**Comment by DavidBoike** [comment]
All the options discussed so far are problematic. I think we should take a different course.

* Ability to customize table name per endpoint: This just creates a bit of secret knowledge you must have to get Outbox to work correctly when you publish events. Even if you knew you had to do it, it would be a really annoying customization to have to do in every single endpoint. Additionally, when you have an organization that disables automatic schema updates, this increases the operational burden when releasing every single endpoint.
* Schema change to include EndpointName column: A schema change would require (right?) a major version bump to 7.0 and be a huge PITA. If we included this in current NH V7, it would only "fix" things for Core V6 and we'd have an awful support burden for Core V5 going forward for years. Alternative is to release NH V7 targeting V5 (with really only that change) and then NH V8 for Core V6. Gross. Plus, this has backwards compatibility issues with "in-flight" outbox records having a null endpoint name.

Instead, let's not touch the schema but instead, fill the MessageId column with `{EndpointName}/{MessageId}` going forward.

The lookup query for outbox records then becomes, assuming an endpoint named OrderService and a MessageId of 1:

`SELECT * FROM OutboxRecord WHERE MessageId IN ('OrderService/1', '1')`

Or with an OR or however NHibernate would do that, in order to run on Oracle too.

Now, in-flight outbox records (MessageId = "1") will be found, as will newer records containing the endpoint name. Note I am talking about the endpoint name, _not the endpoint **instance** name_, so scale-out has nothing to do with this discussion.

This adds a bit of "tribal knowledge" for us, but we can manage that, and more importantly it puts the onus to do so on us rather than on the customer. A major version down the road, we can remove the OR in the SQL for slightly better perf. Or a config option for "I totally have no in-flight outbox records, I would like better perf now, please use new query". That might be a really good option, because it will self-document, and gives us the opportunity to expose/publicize removing the older OR behavior by deprecating the UseFasterQuery() method.

So I propose changing PoA to:

 - [ ] Fill OutboxRecord's message id with `{EndpointName}/{MessageId}`
 - [ ] Query by ID alone or endpoint-qualified ID, ensuring cross-db support
 - [ ] (Optional) Add option to opt-in to query based solely on joint id
 - [ ] Verify if the same problem exists in RavenDB outbox persistence

**Comment by SzymonPobiega** [comment]
I like this approach @DavidBoike !

**Comment by SeanFeldman** [comment]
@SzymonPobiega PoA needs to be updated. Would be good if you could think of potential issues with [this approach](https://github.com/Particular/NServiceBus.NHibernate/issues/117#issuecomment-140213543). We couldn't identify any serious drawbacks.

**Comment by andreasohlund** [comment]
The only thing that comes to mind is the outbox cleaners interfering with
each other (like we saw in the desk case). How can we make sure that each
cleaner only touches the rows it "own" in a safe way? "delete where
dispatched && id.startsWith("MyEndpoint")"

Another option would be to configure NH(in this case) to not care that rows
where deleted underneath it?

**Comment by SzymonPobiega** [comment]
So what we are doing is something like this

```
var result = session.QueryOver<OutboxRecord>().Where(o => o.Dispatched && o.DispatchedAt < dateTime)
    .List();

foreach (var record in result)
{
    session.Delete(record);
}
```

Why not just use `delete` HQL?

**Comment by SeanFeldman** [comment]
@andreasohlund @SzymonPobiega I think you should consider endpoint name as well in the cleanup as optional.

```c#
var result = session.QueryOver<OutboxRecord>().Where(o => o.Dispatched && o.DispatchedAt < dateTime && (o.Id.StartsWith(msgId) || o.Id.Equals(msgId+"\"+endpointName)) )
    .List();
```

**Comment by SzymonPobiega** [comment]
I created a hotfix branch with my changes: https://github.com/Particular/NServiceBus.NHibernate/tree/hotfix-6.2.3

 * Query by id or unique id (`{Endpoint}/{MessageId}`)
 * Set as dispatched by unique id only
 * Store with unique id
 * Delete via `HQL` without checking affected rows

Here's the commit: https://github.com/Particular/NServiceBus.NHibernate/commit/69d0f58d7b1973d36c850076d0ea955d026d9e1e. @andreasohlund can you review before @DavidBoike @colin-higgins and @SeanFeldman wake up?

**Comment by SzymonPobiega** [comment]
@andreasohlund @SeanFeldman here are my thoughts about deletes:
 * The current in memory implementation (load the rows and delete them one-by-one) would cause exceptions not only in multi-endpoint scenario but also in a scale-out scenario where multiple instances of same endpoint compete to delete same rows -> a `HQL` query is needed
 * In order for the `HQL` to work, we need to change the new format to `/{Endpoint}/{MessageId}` and use following query `... WHERE Dispatched = 1 AND DispatchedAt < 'YYY-MM-DD' AND (MessageId LIKE '/Endpoint/%' OR MessageId NOT LIKE '/%')`
 * The above query fetches all rows based on the `DispatchedAt` value and then does in-memory (on SQL Server level) lookup to filter out rows that are not dispatched and owned by other endpoints.
 * Looking at a database with > 4 million rows (simulated 5-day retention outbox @ 10 msg/s) I don't see any performance problems when running these queries

### Query plan

```
StmtText
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  |--Nested Loops(Inner Join, OUTER REFERENCES:([nservicebus].[dbo].[OutboxRecord].[Id], [Expr1002]) OPTIMIZED WITH UNORDERED PREFETCH)
       |--Index Seek(OBJECT:([nservicebus].[dbo].[OutboxRecord].[OutboxRecord_DispatchedAt_Idx]), SEEK:([nservicebus].[dbo].[OutboxRecord].[DispatchedAt] < '2015-09-11 00:00:00.000') ORDERED FORWARD)
       |--Clustered Index Seek(OBJECT:([nservicebus].[dbo].[OutboxRecord].[PK__OutboxRe__3214EC0709B8E1B4]), SEEK:([nservicebus].[dbo].[OutboxRecord].[Id]=[nservicebus].[dbo].[OutboxRecord].[Id]),  WHERE:([nservicebus].[dbo].[OutboxRecord].[Dispatched]=(1)
```

**Comment by johnsimons** [comment]
Sorry for being late to the party!

Looking at the other tables (Subscriptions, Timeouts, ...), we have decided to include the endpoint column, I do understand that is a huge PITA to retrofit this change to the Outbox table but IMO that is the right way to go.
The advantage is that we can index that column (that same way we index that column in the other tables), and save on more expensive queries like `IN` and `LIKE`.
The other reason is that  if we were to start fresh that is what we would do, so I don't see a reason to jeopardise design in name of being a PITA.

**Comment by johnsimons** [comment]
And yes, I did quite a bit of sql profiling to squeeze all performance I could get from SQL and i am super scared of introducing `IN` and `LIKE` into the mix.

**Comment by andreasohlund** [comment]
Regarless if the solution, have we decided if this needs to be backported?

If we don't need to backport I guess @johnsimons has a point in that we
want to build the best solution possible?

**Comment by SzymonPobiega** [comment]
This is a message loss scenario that is happening on any 6.x version on NSB.NHibernate (NSB 5.x). As such, should be addressed on the level of current minor (6.2) as a patch release. No option to add a column here. I am more than happy to discuss pros and cons of adding a column in 7.0 but that would be probably targeting NSB 6.x.

**Comment by andreasohlund** [comment]
Got it, that makes sense. So the scope here is a backwards compatible fix

**Comment by DavidBoike** [comment]
I had forgotten about cleanup, but I don't know why the cleanup really has to employ a `LIKE`. Right now we're doing select + row-wise delete in NHibernate, which I would imagine is wildly less performant than using a LIKE in a single delete statement. If we updated that to `DELETE FROM OutboxRecord WHERE DispatchedAt < @Cutoff` and don't check the results (what's the point?) then we're making performance better than current.

Yeah, we lose the ability to delete just within an endpoint's work if you don't use a LIKE, but is there a lot of point to that? All endpoints sharing a table can cooperate in cleaning it up. The only oddity would be if you had a 7 day expiration policy but wanted to set one endpoint to a 30 day expiration policy for some ridiculous reason. Would someone _do_ that?

Honestly, I don't see any other way forward that can fix 6 and then change schema in 7 and remain forward compatible the whole time.

**Comment by johnsimons** [comment]
@SzymonPobiega 
>As such, should be addressed on the level of current minor (6.2) as a patch release. No option to add a column here.

Why can't we add a column in a patch release ?

**Comment by indualagarsamy** [comment]
I am with @johnsimons. Why is adding a new column problematic in older version? Are we breaking semver if we are still backwards compatible? the old version, just won't use the new column right?

**Comment by SeanFeldman** [comment]
@indualagarsamy are schema changes allowed by customers when upgrading a minor version?

**Comment by johnsimons** [comment]
That is right @indualagarsamy, also I am pretty sure we have done a patch for either timeouts or subscriptions for exactly the same reason, @andreasohlund do u remember ?

**Comment by johnsimons** [comment]
@SeanFeldman the way we recommend upgrading NSB systems is to uninstall old one followed by install new one, or side by side which is the same procedure except reversed.
So the schemas are refreshed as part of the installation.
As far as I know semver looks only at the public face of a programatic API not the data/schema, also adding a column is ok, removing it is where it breaks!

**Comment by DavidBoike** [comment]
And the customers using DisableAutomaticSchemaUpdates() ?

**Comment by indualagarsamy** [comment]
@davidboike - I see that as a communication thing, where we instruct our customers how to get the upgrade, if the case you're talking about is that they don't get the new schema. Deleting a column like @johnsimons mentioned would be breaking.

**Comment by andreasohlund** [comment]
>subscriptions for exactly the same reason, @andreasohlund do u remember ?

Pretty sure we did


>And the customers using DisableAutomaticSchemaUpdates() ?

Can't we code the patch to handle the missing column gracefully, so users that haven't updated the store just gets the old behavior?

In short: I'm with @indualagarsamy and @johnsimons on this one

**Comment by SzymonPobiega** [comment]
I have no strong opinions here. I suspect if we add a `default` clause to the table creation script the old endpoints will be fine. I somehow dislike the fact that in order to upgrade to a minor or patch release we need to run the installers.

**Comment by DavidBoike** [comment]
I was thinking that a change in schema would be a violation of SemVer, at least in spirit if not in letter. If we can add a column in a minor release, and do it transparently as long as you run the installer, and be completely forward compatible, and be sure to over-communicate the schema change, then let's do it.

**Comment by indualagarsamy** [comment]
in the new code, we can detect if the new column ins't there and log a warning to inform the user that there is a new schema is available (especially if they are using `DisableAutomaticSchemaUpdates`) - A warning won't be breaking, plus with the other additional communication on how they patch this, I think that should about cover it?

**Comment by SeanFeldman** [comment]
@SzymonPobiega I've updated PoA to reflect recent discussions as well as added bug-issue mandatory sections.

**Comment by andreasohlund** [comment]
After talking to @johnsimons @udidahan @danielmarbach we came to the conclusion that long term we want to move to a "table"/"collection" per endpoint+storagetype, like we do with the sqlserver queue tables, since this gives us the best isolation but still allows the users to have endpoints sharing the same database.

In the light of this is there an option to instead of doing the column/composite fix we allow the users to specify a specific table per endpoint?

```
config.UsePersistence<NHibernatePersistence>().OutboxTable("billing");

OR

config.UsePersistence<NHibernatePersistence>().UseEndpointNameAsOutboxTableName();
```

**Comment by DavidBoike** [comment]
I feel a little bit like this is running in circles back to the start.

> we came to the conclusion that long term we want to move to a "table"/"collection" per endpoint+storagetype, like we do with the sqlserver queue tables

I'm concerned how to do a transition in that case. And I'm really concerned if the user has to do all this configuration on a per-endpoint basis. With respect to SQL/Oracle, what isolation does separate tables give us that a column and an index doesn't? 

And are we talking going to separate tables for V6-updated or for V7 of NSB.NH?

**Comment by andreasohlund** [comment]
I've had a follow up talk with @indualagarsamy and @SzymonPobiega and we concluded that we should go with a hybrid solution:

1. Make the outbox always safe to use in shared table mode by modifying the ids to contain the endpoint name. This makes sure that no matter what the users does we're safe

2. We also (in the next major) change the persistence to by default use a separate table per endpoint just like we do with sqlserver. ( @SzymonPobiega will raise a separate issue for this)

The reason we decided against the separate column is that it would be confusing when we have no.2 implemented since that would mean a separate table per endpoint but still a column that always has the endpoint name in it.

What do you all think?

**Comment by DavidBoike** [comment]
Two questions:

1. What's the default table naming convention then for NH V7? `{EndpointName}.Outbox` ?
2. What would be the transition plan during a V6 to V7 conversion?

**Comment by andreasohlund** [comment]
Talked with @SimonCropp (who's working on a pure sql server storage sample) and we talked about some thing along the lines of `{endpointname}.{storage type}.storage` to not clash with sqltransport queue tables?

Regarding migration: Yea we'll have to figure that out

**Comment by DavidBoike** [comment]
I can see how Saga storage almost requires NH but it does kind of seem an unnecessary crutch for everything else. And I like `{endpoint}{storagetype}.storage`. And I like getting it fixed with current schema.

And on migration: can't win 'em all I guess. ;-)

**Comment by adamralph** [comment]
@DavidBoike are you task lead here? should you be assigned?

**Comment by DavidBoike** [comment]
Sure I can be task lead. Updated PoA for composite key with current schema + straight SQL cleanup.

If there are no objections I assume this is ready to go, and I'll probably start work on it today or tomorrow.

If anyone else would like to jump in the first 2 PoA points could be done in parallel.

**Comment by andreasohlund** [comment]
>f there are no objections I assume this is ready to go, and I'll probably start work on it today or tomorrow.

We need to craft the customer communication and perhaps flesh out the impact (ie: https://github.com/Particular/PlatformDevelopment/blob/master/IssueTemplate.md)

**Comment by andreasohlund** [comment]
I was thinking about the Customer Communication for this one and realized

1. The scope of this issues is only the non-message driven transports (ASB + Rabbit) since v5 still makes the message ids unique - https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Unicast/Publishing/StorageDrivenPublisher.cs#L46  

2. We need to tell the users how to recover from this situation. Can they use SI to find the events that where effected? Ie filter on the problematic events (since one of the endpoint would have processed it) and then fetch the body and send them straigt to the endpoints that missed them?

**Comment by DavidBoike** [comment]
Ah, you're saying if you're using MSMQ or SQL transport and publish an event, because there are no topics, each of those messages gets a unique ID and then Outbox is happy. Good - I would think that decreases the surface area of this bug pretty significantly.

Best we could do for recovery would be to dredge SC info for events, then analyze the processing handlers, essentially doing a GroupBy on those. Anything with multiple possible handlers, then in each case there may have been one or more missed handlers.

However, it would still be possible that due to chance, or maybe one handler processes quite a bit more quickly, that the same handler was _always_ the first to process any given message, and _never_ processed on the slower/unlucky handler. An automated tool would have no way to distinguish that from an event that has just one rightful subscriber.

Should we add creation of such a tool to the PoA?

**Comment by DavidBoike** [comment]
Updated bug sections to be more explicit and more customer communication appropriate, along with generating a customer communication.

**Comment by andreasohlund** [comment]
>Should we add creation of such a tool to the PoA?


Since this is message loss I don't think we have a choice. 


I wish we had a more powerfull query api in SI to avoid building custom tools for this every time :(

```
select * from audit 
where ProcessingEndpoint in ("EndpointXSharingOutboxDB","EndpointYSharingOutboxDB") and 
MessageType Is Event
```

**Comment by DavidBoike** [comment]
We just agreed in Platform Dev sync that a full-fledged tool is probably overkill. It's a distinct possibility we'd spend a lot of time on it and then nobody would use it, and it certainly couldn't do anything automatic to fix anything.

Instead, we can develop a Raven Studio query that could be run in ServiceControl in maintenance mode and provide instructions for executing it. This would let the customer know they might have an issue and instruct them to contact us at support and we'll help them work through it.

I added the Raven query to the PoA and I will start coding on this on Thursday.

**Comment by DavidBoike** [comment]
First two points of PoA addressed in PR #122

**Comment by DavidBoike** [comment]
Here's the details for the Raven query:

## Index Definition
I named mine **FindOutboxMissedMessages** which is only important if you use the URLs I'll add later.

### Map

```
from msg in docs.ProcessedMessages
where msg.MessageMetadata.MessageIntent == 2
select new 
{
	MsgType = msg.MessageMetadata.MessageType,
	Endpoints = new string[] { msg.MessageMetadata.ReceivingEndpoint.Name },
	MsgCount = 1,
	EndpointCount = 1
}
```

### Reduce
```
from result in results
group result by result.MsgType into g
select new
{
	MsgType = g.Key,
	Endpoints = g.SelectMany(x => x.Endpoints).Distinct().ToArray(),
	MsgCount = g.Sum(x => x.MsgCount),
	EndpointCount = g.SelectMany(x => x.Endpoints).Distinct().Count()
}
```
### Fields
* EndpointCount, Indexing = Analyzed

## Queries
Executing an empty query against this index ([equivalent direct db query](http://localhost:33333/storage/indexes/FindOutboxMissedMessages)) will return data for all events. This query can actually be executed even if you don't add the field definition.

![screenshot 2015-10-02 11 57 08](https://cloud.githubusercontent.com/assets/427110/10252379/b92a111e-68fc-11e5-857f-53a2e2495fec.png)

A query of `EndpointCount:[2 TO *]` ([equivalent direct db query](http://localhost:33333/storage/indexes/FindOutboxMissedMessages?query=EndpointCount:[2%20TO%20*])) will show only events that have been processed, at some point, by more than one endpoint, which should really be the topic of scrutiny. This one *requires* the field definition otherwise you won't be able to query on EndpointCount.

![screenshot 2015-10-02 11 55 54](https://cloud.githubusercontent.com/assets/427110/10252360/9be4911a-68fc-11e5-82cf-222a1b7c7076.png)

I didn't spin up Rabbit or anything to really reproduce the error, this is just using MSMQ and having one event having 1 subscriber and one having 2. So really, it just shows multi-subscriber events, it doesn't reach back in and analyze individual messages to be sure they were processed by ALL endpoints. Not sure that's going to be possible with Raven without a more complex tool, but I'm going to try that next.

So the upshot is that a customer communication would need to be very specific that you shouldn't even run this tool unless the other requirements (NH, Outbox, Rabbit/ASB) are met, because otherwise it may prompt you to be concerned about a system that has no problems.

**Comment by DavidBoike** [comment]
To spell out which messages may be affected required a grand total of 2 Map/Reduce indexes and some C# glue to bring them together.

This is the result:
https://github.com/DavidBoike/NHibernate.Issue117Detector

Definitely in the PoC stage and not polished by any means.

**Comment by DavidBoike** [comment]
Example output on my machine:

```
MissedEvents.Messages.MultiSubscriberEvent handled by 2 endpoints:
  Endpoints:
    * MissedEvents.SubscriberA
    * MissedEvents.SubscriberB
  Problem Messages:
    * Msg 236d2376-ddcf-4bc2-9567-a52500ae3649 not processed by 1 endpoints:
      MissedEvents.SubscriberB
    * Msg 98581215-decb-47d0-b43e-a52500ae3557 not processed by 1 endpoints:
      MissedEvents.SubscriberB
    * Msg 895b97c8-b610-415b-b1de-a52500ae2fb5 not processed by 1 endpoints:
      MissedEvents.SubscriberB
    * Msg 534d0235-3b86-4f12-a41c-a52500ae3608 not processed by 1 endpoints:
      MissedEvents.SubscriberB
    * Msg 1b80c0da-ea1d-40c0-aeec-a52500e84cab not processed by 1 endpoints:
      MissedEvents.SubscriberA
    * Msg 0fa700f1-bac1-4153-a585-a52500e84fc0 not processed by 1 endpoints:
      MissedEvents.SubscriberA
    * Msg ba8314f6-f6f6-4c2d-ae4b-a52500e8e0d9 not processed by 1 endpoints:
      MissedEvents.SubscriberA
    * Msg dfa0b51e-ce4d-41f9-86bc-a52500e8e580 not processed by 1 endpoints:
      MissedEvents.SubscriberA
Complete
```

This data is just me messing around with MSMQ though, by having in-memory persistence and not starting one endpoint so that an event only goes to one subscriber. This tells me that **if you're using MSMQ with Outbox enabled, you may be hit by this bug too**. It looks like the logical message ID, which NServiceBus generates (not the MSMQ infrastructure id) is the same for multiple sent copies of a published message. Am I off base?

**Comment by andreasohlund** [comment]
>if you're using MSMQ with Outbox enabled, you may be hit by this bug too

Yes but only if you have InMemory persistence for your subscriptions right?

On Fri, Oct 2, 2015 at 10:11 PM, David Boike <notifications@github.com>
wrote:
>
> Example output on my machine:
>
> MissedEvents.Messages.MultiSubscriberEvent handled by 2 endpoints:
>   Endpoints:
>     * MissedEvents.SubscriberA
>     * MissedEvents.SubscriberB
>   Problem Messages:
>     * Msg 236d2376-ddcf-4bc2-9567-a52500ae3649 not processed by 1
endpoints:
>       MissedEvents.SubscriberB
>     * Msg 98581215-decb-47d0-b43e-a52500ae3557 not processed by 1
endpoints:
>       MissedEvents.SubscriberB
>     * Msg 895b97c8-b610-415b-b1de-a52500ae2fb5 not processed by 1
endpoints:
>       MissedEvents.SubscriberB
>     * Msg 534d0235-3b86-4f12-a41c-a52500ae3608 not processed by 1
endpoints:
>       MissedEvents.SubscriberB
>     * Msg 1b80c0da-ea1d-40c0-aeec-a52500e84cab not processed by 1
endpoints:
>       MissedEvents.SubscriberA
>     * Msg 0fa700f1-bac1-4153-a585-a52500e84fc0 not processed by 1
endpoints:
>       MissedEvents.SubscriberA
>     * Msg ba8314f6-f6f6-4c2d-ae4b-a52500e8e0d9 not processed by 1
endpoints:
>       MissedEvents.SubscriberA
>     * Msg dfa0b51e-ce4d-41f9-86bc-a52500e8e580 not processed by 1
endpoints:
>       MissedEvents.SubscriberA
> Complete
>
> This data is just me messing around with MSMQ though, by having in-memory
persistence and not starting one endpoint so that an event only goes to one
subscriber. This tells me that if you're using MSMQ with Outbox enabled,
you may be hit by this bug too. It looks like the logical message ID, which
NServiceBus generates (not the MSMQ infrastructure id) is the same for
multiple sent copies of a published message. Am I off base?

**Comment by DavidBoike** [comment]
I don't think so. If you have SQL Server with AlwaysOn, you have no DTC support, and so you could use MSMQ with no DTC, and enable Outbox with NHibernate. And then, because the logical, NSB-generated message id is the same, even though it's storage-driven publishing, you still have the same problem.

Right?

**Comment by andreasohlund** [comment]
Hmm, SqlServer is using the messagedriven pubsub so it shouldn't be affected? (unless you're testing this on core v6)

**Comment by DavidBoike** [comment]
Confirmed: Msmq + Disable DTC + NH Outbox causes the bug. Going to test SQL Server transport next.

**Comment by DavidBoike** [comment]
Problem with SqlServer transport too. I think we need to assume it's bad on all transports. Will update issue text accordingly.

**Comment by DavidBoike** [comment]
Although this case was closed to generate release notes, the PoA is now completed and I am calling it done-done.
